charmcraft/jujuignore.py

```diff
@@ -0,0 +1,235 @@
+# Copyright 2020 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# For further info, check https://github.com/canonical/charmcraft
+
+import logging
+import re
+import typing
+
+logger = logging.getLogger(__name__)
+
+KEEP = 'keep'
+SKIP = 'skip'
+FORCEKEEP = 'forcekeep'
+
+
+_unescapes = {
+    r'\!': '!',
+    r'\ ': ' ',
+    r'\#': '#',
+
+}
+
+
+def _rstrip_unescaped(rule):
+    """Remove trailing whitespace that isn't escaped."""
+    i = len(rule) - 1
+    while i >= 0:
+        if rule[i] == '\n' or rule[i] == '\r':
+            rule = rule[:i]
+        elif rule[i] != ' ':
+            break
+        elif i == 0 or rule[i - 1] != '\\':
```

Comment:
it can, the while loop is i >= 0. 
If it is 0 the code 'works' because rule[-1] is the last character, but that feels like it 'works' for the wrong reason.

Reply:
ack